components/tidb_query/src/expr/builtin_string.rs

```diff
@@ -914,6 +914,42 @@ impl ScalarFunc {
             .map(|i| 1 + i as i64)
             .or(Some(0)))
     }
+
+    // See https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_quote
+    pub fn quote<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s = try_opt_or!(
+            self.children[0].eval_string(ctx, row),
+            Some(Cow::Borrowed(b"NULL"))
+        );
+        let mut count = 2;
+        for byte in s.iter() {
+            if *byte == b'\'' || *byte == b'\\' || *byte == b'\0' || *byte == 26u8 {
+                count += 1;
+            }
+        }
```

Comment:
I just did a quick benchmark. The two-round solution outperforms the maybe-reallocate one only if the string is very short (when iterating the whole string is cheaper than one allocation).

Reply:
I would prefer to simply `s.len() * 2` so that there is no need to iterate the `s` first.. What do you think? @sticnarf

Reply:
Good idea.

Reply:
I guess it would make more sense to use `s.len() * 2 + 2` since it would make sure no reallocation regarding the input stirng.